packages/teleterm/src/preload.ts

```diff
@@ -7,28 +7,38 @@ import { ElectronGlobals } from './types';
 import { createPtyService } from 'teleterm/services/pty/ptyService';
 import { readGrpcCerts } from './services/grpcCerts';
 
-async function startPreload(): Promise<void> {
-  const mainProcessClient = createMainProcessClient();
-  const runtimeSettings = mainProcessClient.getRuntimeSettings();
-  const loggerService = createLoggerService({
-    dev: runtimeSettings.dev,
-    dir: runtimeSettings.userDataDir,
-    name: 'renderer',
-  });
+const mainProcessClient = createMainProcessClient();
+const runtimeSettings = mainProcessClient.getRuntimeSettings();
+const loggerService = createLoggerService({
+  dev: runtimeSettings.dev,
+  dir: runtimeSettings.userDataDir,
+  name: 'renderer',
+});
 
-  PreloadLogger.init(loggerService);
+PreloadLogger.init(loggerService);
 
-  const grpcCerts = await readGrpcCerts(runtimeSettings.certsDir);
+contextBridge.exposeInMainWorld('electron', getElectronGlobals());
 
-  const tshClient = createTshClient(runtimeSettings, grpcCerts);
-  const ptyServiceClient = createPtyService(runtimeSettings, grpcCerts);
+async function getElectronGlobals(): Promise<ElectronGlobals> {
+  try {
+    const [grpcCerts, resolvedAddresses] = await Promise.all([
+      readGrpcCerts(runtimeSettings.certsDir),
+      mainProcessClient.getResolvedChildProcessAddresses(),
```

Comment:
How do we guarantee that these address are set before the execution gets to this point?

Reply:
Please take a look at `_init()` method in `mainProcess.ts`. The promise that resolves that addresses is set in that method. This means that by the time the execution reaches that point, the promise is already created and our async handler will return it.

To improve readability, I extracted the code that sets ` this.resolvedChildProcessAddresses...` to a new method and changed the order of execution. Now `this._initIpc()` is set last.

Reply:
Great, that makes sense, thanks.